### ISSUE TITLE
"Callable | Callable" not callable [misc]

### ISSUE THREAD

**Issue by jolaf** [python/mypy#17631]
The following code:
```python3
from collections.abc import Callable, Sequence
from functools import partial, reduce
from typing import cast, Any

CA = Callable[[Any], Any]
CB = Callable[[Callable[..., Any], Any], Any]

def f(s: Sequence[CA | CB]) -> CA:
    return cast(CA, reduce(lambda a, b: partial(b, a), s))
```
produces the following output:
```shell
test.py: note: In function "f":
test.py:9:41: error: "Callable[[Any], Any] | Callable[[Callable[..., Any], Any], Any]" not callable  [misc]
        return cast(CA, reduce(lambda a, b: partial(b, a), s))
                                            ^
```
This looks like a false positive to me.
```shell
$ mypy --version
mypy 1.11.1 (compiled: yes)

$ python --version
Python 3.12.3

$ lsb_release -d
No LSB modules are available.
Description:	Ubuntu 24.04 LTS
```

**Comment by hauntsaninja** [comment]
This is the same as many other issues, where mypy computes suboptimal join between callables. In this case join of `[def (Any) -> Any` and `def (def (*Any, **Any) -> Any, Any) -> Any]`. Happens to shows up in 1.11 because we now type check functools.partial more closely, but underlying issue is old. You can also see this with:
```
CA = Callable[[Any], Any]
CB = Callable[[Callable[..., Any], Any], Any]

def f(a: CA, b: CB):
    reveal_type([a, b])  # list[builtins.function], should be list[Callable[[Any], Any]] or something
```

**Comment by bswck** [comment]
I've taken on the challenge of fixing this. It's my first time working with mypy, so bear with me or let me know if someone else fixes it.

**Comment by jolaf** [comment]
> I've taken on the challenge of fixing this. It's my first time working with mypy, so bear with me or let me know if someone else fixes it.

Thanks!!

**Comment by hauntsaninja** [comment]
This was fixed in #17903
